mitodl/ocw-hugo-themes
# Clean up env variables + make them all optional

### Pre-Flight checklist

  - [x] Code is tested *Insofar as tests still pass, and no new application functionality was added.*

### What are the relevant tickets?
This PR is a followup to #941 in regards to #919 .

Also closes #962 

### What's this PR do?
This PR reoganizes how we access environment variables in development scripts. When working on #941 it became clear that having environment variables to control things like webpack's host+port is convenient, but also that we did not have a great way to set default values for environment variables.

This PR:
1. Introduces a `env.ts`, analogous to `settings.py` in Django. It does simple validation of env variable values, sets defaults, and provides useful error messages for things not set. In particular, because `ocw-hugo-themes` requires no secret information (e.g., API keys), *all environment variables are now optional.*
2. Adds a CLI command `with-env [options] <command...>` that runs a command with the values of `env.ts`.
3. Enables setting webpack's host via an environment variable
4. Expands linting a little bit. *I added a few new files that would not otherwise have been linted.*
5. Fixes an issue with stopping `yarn start:course`. And adds better color / labeling. 

⚠️ For code review, I mildly suggest looking [commit-by-commit](https://github.com/mitodl/ocw-hugo-themes/pull/964/commits)

### How should this be manually tested?

⚠️ First, check out [cc/webpack_host](https://github.com/mitodl/ocw-hugo-projects/pull/229) in OCW Hugo Projects.

#### Linting/formatting changes:

Below, introduce the error in any js/ts/scss file outside of `dist` folders.

1. Introduce a linting error (e.g., unused import). `yarn lint` should throw and error.
2. Introduce a formatting error (e.g., extra space. Depending on editor, may need to "save without formatting". Run `yarn fmt:check` followed by `echo process exited with code $?`. You should see that the process exited with code 1.

#### Stopping `start:course`.

#941 introduced a minor bug where dev processes like `start:course` would not stop correctly. If you try to stop `start:course` on `main`, you will find that the terminal output does not fully quit. You have to (1)`ctrl + c`, (2) wait a sec, (3) hit any other key. This PR removes the need for (3).

1. Run `yarn start:course`, and wait for the site to go live.
    - **New colors / labels**: The output of `yarn start:course`, etc, should now have colored labels "[webpack]" or "[hugo]".
2. Hit ctrl+c to stop the servers.
8. Wait a few seconds, and hit any key arrow. *Do you get the behavior you would expect? Does "hitting a key" introduce any extra escape sequences?*

#### Env variable changes
1. Comment out your entire env file. Run commands `yarn start:www`, `yarn start:course`, and `yarn start:www`. They should all work. *Although, there are some assumptions, like `ocw-hugo-projects` and `ocw-content-prod` are siblings of your `ocw-hugo-themes`*
2. Un-comment your env file, and check that the `yarn start:thing` commands all use your env values.
3. Check that the `start:thing` commands behave as expected with inline overrides, e.g.,  `OCW_TEST_COURSE=your-favorite-course yarn start:thing`.
9. Check that the [steps in readme](https://github.com/mitodl/ocw-hugo-themes/tree/cc/env-cleanup-2#environment-variables) for seeding a `.env` work correctly.
10. Check that `yarn build /path/to/config /path/to/course` still works. *Alert! You'll need a fully-fledged env file to run this; this command assumes that you want production mode, so fewer default values are set.*
11. Remove `STATIC_API_BASE_URL` from your env file and check that `yarn build /path/to/config /path/to/course` provides useful error messages when some env variables are missing.

#### Check new WEBPACK_HOST environment variable

Set up ocw-hugo-themes to serve other devices on your local network.

1. Determine your ~public~ local[^1] IPv4 address. This varies by OS. You can try asking [what is my ip address](https://www.google.com/search?q=what+is+my+ip+address&sxsrf=ALiCzsb1O05dsi4ZST9oxeKJxtzH70sg0w%3A1667859021345&ei=TYJpY7bMFPyy5NoPpJmlkAM&ved=0ahUKEwj2poKEi537AhV8GVkFHaRMCTIQ4dUDCBA&uact=5&oq=what+is+my+ip+address&gs_lcp=Cgxnd3Mtd2l6LXNlcnAQAzIFCAAQgAQyBQgAEIAEMgUIABCABDIFCAAQgAQyBQgAEIAEMgUIABCABDIFCAAQgAQyBQgAEIAEMgUIABCABDIFCAAQgAQ6CggAEEcQ1gQQsAM6BwgAELADEEM6DQgAEOQCENYEELADGAE6EgguEMcBENEDEMgDELADEEMYAjoECCMQJzoFCAAQkQJKBAhNGAFKBAhBGABKBAhGGAFQhgRYxw1g-Q5oAXABeACAAVaIAfkFkgECMTCYAQCgAQHIARHAAQHaAQYIARABGAnaAQYIAhABGAg&sclient=gws-wiz-serp) *Although, for me Google will only tell me my IPv6, and I had trouble with this when using my IPv6 address. Probably can work around it with some escaping or something... but IPv4 is simple.*
2. Set `WEBPACK_HOST=<your local IPv4 address>`
3. `yarn start:course`
4. Visit `http://your_local_ip_address:3000`. It should work.
6. Visit `http://your_local_ip_address:3000` **on a different device** within the same wifi network. It should work, including loading the webpack assets. 

[^1]: https://github.com/mitodl/ocw-hugo-themes/pull/964#pullrequestreview-1174273854
